src/btree/bt_walk.c

```diff
@@ -17,54 +17,52 @@ __ref_index_slot(WT_SESSION_IMPL *session,
-	/*
-	 * Copy the parent page's index value: the page can split at any time,
-	 * but the index's value is always valid, even if it's not up-to-date.
-	 */
-retry:	WT_INTL_INDEX_GET(session, ref->home, pindex);
+	for (;;) {
+		/*
+		 * Copy the parent page's index value: the page can split at
+		 * any time, but the index's value is always valid, even if
+		 * it's not up-to-date.
+		 */
+		WT_INTL_INDEX_GET(session, ref->home, pindex);
+		entries = pindex->entries;
 
-	/*
-	 * Use the page's reference hint: it should be correct unless the page
-	 * split before our slot.  If the page splits after our slot, the hint
-	 * will point earlier in the array than our actual slot, so the first
-	 * loop is from the hint to the end of the list, and the second loop
-	 * is from the start of the list to the end of the list.  (The second
-	 * loop overlaps the first, but that only happen in cases where we've
-	 * split the tree and aren't going to find our slot at all, that's not
-	 * worth optimizing.)
-	 *
-	 * It's not an error for the reference hint to be wrong, it just means
-	 * the first retrieval (which sets the hint for subsequent retrievals),
-	 * is slower.
-	 */
-	i = ref->pindex_hint;
-	if (i < pindex->entries && pindex->index[i] == ref) {
-		*pindexp = pindex;
-		*slotp = i;
-		return;
-	}
-	while (++i < pindex->entries)
-		if (pindex->index[i] == ref) {
-			*pindexp = pindex;
-			*slotp = ref->pindex_hint = i;
-			return;
-		}
-	for (i = 0; i < pindex->entries; ++i)
-		if (pindex->index[i] == ref) {
-			*pindexp = pindex;
-			*slotp = ref->pindex_hint = i;
-			return;
+		/*
+		 * Use the page's reference hint: it should be correct unless
+		 * there was a split or delete in the parent before our slot.
+		 * If the hint is wrong, it can be either too big or too small,
+		 * but often only by a small amount.  Search up and down the
+		 * index starting from the hint.
+		 *
+		 * It's not an error for the reference hint to be wrong, it
+		 * just means the first retrieval (which sets the hint for
+		 * subsequent retrievals), is slower.
+		 */
+		WT_ORDERED_READ(hint, ref->pindex_hint);
+		if ((i = hint) < entries && pindex->index[i] == ref)
+			goto found;
+		maxslots = hint > entries / 2 ? hint : entries - hint;
+		for (j = 1; j <= maxslots; j++) {
+			if ((i = hint + j) < entries && pindex->index[i] == ref)
+				goto found;
+			if (j <= hint &&
+			    (i = hint - j) < entries && pindex->index[i] == ref)
+				goto found;
 		}
 
-	/*
-	 * If we don't find our reference, the page split and our home pointer
-	 * references the wrong page. When internal pages split, their WT_REF
-	 * structure home values are updated; yield and wait for that to happen.
-	 */
-	__wt_yield();
-	goto retry;
+		/*
+		 * If we don't find our reference, the page split and our home
+		 * pointer references the wrong page. When internal pages
+		 * split, their WT_REF structure home values are updated; yield
+		 * and wait for that to happen.
+		 */
+		__wt_yield();
+	}
```

Comment:
OK... I just figured out why the code moving to smaller slots has to test against `entries`, but I think that also means that if there's a big number in the hint field, we're potentially going to be in this loop for a long time.